File: src/main/java/cz/vhromada/catalog/facade/SongFacade.java
package cz.vhromada.catalog.facade;

import cz.vhromada.catalog.entity.Music;
import cz.vhromada.catalog.entity.Song;
import cz.vhromada.result.Result;

/**
 * An interface represents facade for songs.
 *
 * @author Vladimir Hromada
 */
public interface SongFacade extends CatalogChildFacade<Song, Music> {

    /**
     * Adds song. Sets new ID and position.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Music is null</li>
     * <li>Music ID is null</li>
     * <li>Music doesn't exist in data storage</li>
     * <li>Song is null</li>
     * <li>Song ID isn't null</li>
     * <li>Name is null</li>
     * <li>Name is empty string</li>
     * <li>Length of song is negative value</li>
     * <li>Note is null</li>
     * </ul>
     *
     * @param parent music
     * @param data  song
     * @return result with validation errors
     */
    @Override
    Result<Void> add(Music parent, Song data);

    /**
     * Updates song.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Song is null</li>
     * <li>ID isn't null</li>
     * <li>Name is null</li>
     * <li>Name is empty string</li>
     * <li>Length of song is negative value</li>
     * <li>Note is null</li>
     * <li>Song doesn't exist in data storage</li>
     * </ul>
     *
     * @param data new value of song
     * @return result with validation errors
     */
    @Override
    Result<Void> update(Song data);

}

File: src/main/kotlin/cz/vhromada/catalog/facade/SongFacade.kt
package cz.vhromada.catalog.facade

import cz.vhromada.catalog.entity.Music
import cz.vhromada.catalog.entity.Song
import cz.vhromada.common.facade.MovableChildFacade
import cz.vhromada.validation.result.Result

/**
 * An interface represents facade for songs.
 *
 * @author Vladimir Hromada
 */
interface SongFacade : MovableChildFacade<Song, Music> {

    /**
     * Adds song. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * Music ID is null
     *  * Music doesn't exist in data storage
     *  * Song ID isn't null
     *  * Song position isn't null
     *  * Name is null
     *  * Name is empty string
     *  * Length of song is null
     *  * Length of song is negative value
     *  * Note is null
     *
     * @param parent music
     * @param data   song
     * @return result with validation errors
     */
    override fun add(parent: Music, data: Song): Result<Unit>

    /**
     * Updates song.
     * <br></br>
     * Validation errors:
     *
     *  * ID isn't null
     *  * Position is null
     *  * Name is null
     *  * Name is empty string
     *  * Length of song is null
     *  * Length of song is negative value
     *  * Note is null
     *  * Song doesn't exist in data storage
     *
     * @param data new value of song
     * @return result with validation errors
     */
    override fun update(data: Song): Result<Unit>

}
